crates/red_knot/src/salsa_db/semantic/types/infer.rs

```diff
@@ -0,0 +1,396 @@
+    file: File,
+    enclosing_scope: ScopeId,
+
+    symbol_table: Arc<SymbolTable>,
+    control_flow_graph: Arc<FlowGraph>,
+    typing_scopes: &'a TypingScopes,
+
+    // TODO: This is going to be somewhat slow because we need to map the AST node to the expression id for
+    //   every expression in the body. That's a lot of hash map lookups.
+    //  We can't use an `IndexVec` here because a) expression ids are per module and b) the type inference
+    //  builder visits the expressions in evaluation order and not in pre-order.
+    result: TypeInference,
+}
+
+impl<'a> TypeInferenceBuilder<'a> {
+    fn new(db: &'a dyn Db, scope: TypingScope) -> Self {
+        let file = scope.file(db);
+        let symbol_table = symbol_table(db, file);
+        let control_flow_graph = flow_graph(db, file);
+
+        let symbol_scope = match scope {
+            TypingScope::Function(function) => {
+                symbol_table.scope_id_for_node(NodeWithScopeId::Function(function.id(db)))
+            }
+            TypingScope::Class(class) => {
+                symbol_table.scope_id_for_node(NodeWithScopeId::Class(class.id(db)))
+            }
+            TypingScope::Module(file) => SymbolTable::root_scope_id(),
+        };
+
+        let builder = TypeInferenceBuilder {
+            db,
+            typing_scope: scope,
+            file,
+            enclosing_scope: symbol_scope,
+
+            symbol_table,
+            control_flow_graph,
+            typing_scopes: typing_scopes(db, file),
+
+            result: TypeInference::default(),
+        };
+        builder
+    }
+
+    fn typing_context(&self) -> TypingContext {
+        TypingContext::local(self.db, self.typing_scope, &self.result)
+    }
+
+    fn lower_module(&mut self, module: &ModModule) {
+        self.result.module = Some(ModuleType::new(self.typing_scope.file(self.db)));
+
+        self.visit_body(&module.body);
+    }
+
+    fn lower_class_body(&mut self, class: &StmtClassDef) {
+        self.visit_body(&class.body);
+    }
+
+    fn lower_function_body(&mut self, function: &StmtFunctionDef) {
+        self.visit_body(&function.body);
+    }
+
+    fn lower_import(&mut self, import: &StmtImport) {
+        let import_id = import.ast_id(self.db, self.file);
+
+        for (i, name) in import.names.iter().enumerate() {
+            let ty = if let Some(module) = resolve_module_name(self.db, ModuleName::new(&name.name))
+            {
+                // TODO: It feels hacky to resolve the module ID out of the blue like this.
+                Type::Module(GlobalTypeId::new(
+                    TypingScope::Module(module.file()),
+                    ModuleTypeId,
+                ))
+            } else {
+                Type::Unknown
+            };
+
+            self.result.local_definitions.insert(
+                ImportDefinition {
+                    import: import_id,
+                    name: u32::try_from(i).unwrap(),
+                }
```

Comment:
Is your concern just about the `ImportDefinition` creation that is used as key? Because there's a difference. We associate a definition with its type. 

I don't think we can avoid this much without having a way to iterate over the AST and definitions at the same time.